components/debounce/browser/debounce_rule.cc

```diff
@@ -86,15 +143,44 @@ bool DebounceRule::Apply(const GURL& original_url, GURL* final_url) const {
   if (!include_pattern_set_.MatchesURL(original_url))
     return false;
 
-  std::string unescaped_value;
-  if (!net::GetValueForKeyInQuery(original_url, param_, &unescaped_value))
-    return false;
-  if ((action_ == kDebounceBase64DecodeAndRedirectToParam) &&
-      (!base::Base64UrlDecode(unescaped_value,
-                              base::Base64UrlDecodePolicy::IGNORE_PADDING,
-                              &unescaped_value))) {
+  if (!DebounceRule::CheckPrefForRule(prefs)) {
     return false;
   }
+
+  std::string unescaped_value;
+
+  if (action_ == kDebounceRegex) {
+    // apply param regex to original url.
+    const re2::RE2 kPatternRegex(param_);
+    if (!kPatternRegex.ok()) {
+      LOG(WARNING) << "Debounce rule has param: " << param_
+                   << " which is an invalid regex pattern";
+      return false;
+    }
+    if (kPatternRegex.NumberOfCapturingGroups() != 1) {
+      LOG(WARNING) << "Debounce rule has param: " << param_
+                   << " which captures != 1 groups";
+      return false;
+    }
+    if (!RE2::PartialMatch(original_url.spec(), kPatternRegex,
```

Comment:
What is the maximum regex size/complexity? Is it okay to parse & match it on UI thread?

Reply:
We manually add the regex and each one will be reviewed. Other than that, I don't have ideas on how to bound the complexity here.

Reply:
If we don't have any estimations/limitations, I suggest moving this code away from UI thread.

Anyway, the total checking time (= linear matching agains all the rules) should not exceed ~10-20ms, because this code blocks each navigations.

Reply:
I see that there's a way to bind the max memory used for the regex and a rough estimate of the complexity can be given by `ProgramSize`. I can come up with some examples that this feature would use and get rough bounds for memory and complexity.

Reply:
> because this code blocks each navigations.

Just FYI, this code will not block every navigation - it only gets run if the MatchPattern in the `includes` matches. So we're only applying this regex on suspected "bad" URLs.
Also, re2 is linear time and doesn't support backtracking, lookaheads etc. I suggest just capping the length of the regex as a safeguard.

Reply:
Added a max memory check per regex and bound the length of the regex pattern.